### PR TITLE
Fix QLoRA recipe, add CI

### PR DESCRIPTION
**Pull request by rohan-varma**
#### Context
- As reported by https://github.com/pytorch/torchtune/issues/570 and by @ebsmothers, QLoRA recently broke due to torchao nightly update. In this PR, I pin torchao-nightly to the last known working version which contains all features we need, re-enable the unittests, and add a recipe test for QLoRA.

#### Changelog
- See above
- Also a couple of changes to `get_precision`. In particular, we accept a `device` argument which is mostly used for bf16 support checking. bf16 can run on CPU (there might be some unsupported ops versus CUDA in general), so I don't think we should run in fp32 for CPU if bf16 is specified.
- Basically, previously if bf16 was specified, but the GPU did not support, we'd return torch.float32, even if the user was training on CPUs. Now, if the GPU does not support but the user is using CPU, we return bf16.

#### Test plan
- Unit:  pytest tests/recipes/test_lora_finetune_single_device.py -v -k test_loss
- Fresh install, verify expected ao nightly. Run QLoRA: `tune lora_finetune_single_device --config recipes/configs/qlora_finetune_single_device.yaml`

#### Follow up
- We need to establish a process for regularly upgrading deps for libraries which we use nightlies for. The simplest thing to do is what I've done now - pin the version and bump it whenever a new feature/bugfix we need is released. Though if there's a large gap in time between these, the upgrade process could be time consuming and cumbersome to resolve all issues. I could see a process where we upgrade ~weekly to avoid this and more regularly deal with the issues. Unfortunately since we're using nightlies of a library that doesn't provide BC guarantee at the moment, its tricky to think of a more robust upgrade process.

### CODE REVIEW

**Comment by pbontrager** [comment]
> We need to establish a process for regularly upgrading deps for libraries which we use nightlies for. The simplest thing to do is what I've done now - pin the version and bump it whenever a new feature/bugfix we need is released. Though if there's a large gap in time between these, the upgrade process could be time consuming and cumbersome to resolve all issues. I could see a process where we upgrade ~weekly to avoid this and more regularly deal with the issues. Unfortunately since we're using nightlies of a library that doesn't provide BC guarantee at the moment, its tricky to think of a more robust upgrade process.

For MMF we use a CI bot that automatically creates a PR to update the pinned version whenever the dependency is updated. This allows it to be tested before it's merged into main while keeping the dependencies up to date at the same time.

File: tests/recipes/test_lora_finetune_single_device.py [hunk 1]
```diff
@@ -43,19 +43,24 @@ def _get_test_config_overrides(self):
...
+    def _fetch_expected_loss_values(self, run_qlora: bool = False):
+        if run_qlora:
+            return [10.5057, 10.5575, 10.5179, 10.4898]
+        else:
+            return [10.5074, 10.5614, 10.5205, 10.4918]
 
     @pytest.mark.integration_test
-    def test_loss(self, tmpdir, monkeypatch):
+    @pytest.mark.parametrize("run_qlora", [True, False])
+    def test_loss(self, run_qlora, tmpdir, monkeypatch):
```
Comment (ebsmothers) [inline tests/recipes/test_lora_finetune_single_device.py:54]: Given that the loss, config file, and overrides are all different, maybe this just merits a separate `test_qlora_loss` test case? I know it adds copypasta but I think it'd be clearer

**Review by joecummings — APPROVED** [review]
go go go
